packages/libs/error-stack/src/fmt/hook.rs

```diff
@@ -0,0 +1,652 @@
+    /// This corresponds to the output of [`std::fmt::Formatter::alternate`].
+    #[must_use]
+    pub fn alternate(&self) -> bool {
+        self.parent.alternate
+    }
+}
+
+impl<T: 'static> HookContext<'_, T> {
+    /// Return a reference to a value of type `U`, if a value of that type exists.
+    ///
+    /// Values returned are isolated and therefore "bound" to `T`, this means that if two different
+    /// [`HookContext`]s that share the same inner value (e.g. same invocation of [`Debug`]) will
+    /// return the same value.
+    ///
+    /// [`Debug`]: core::fmt::Debug
+    #[must_use]
+    pub fn get<U: 'static>(&self) -> Option<&U> {
+        self.parent
+            .storage
+            .get(&TypeId::of::<T>())?
+            .get(&TypeId::of::<U>())?
+            .downcast_ref()
+    }
+
+    /// Return a mutable reference to a value of type `U`, if a value of that type exists.
+    ///
+    /// Values returned are isolated and therefore "bound" to `T`, this means that if two different
+    /// [`HookContext`]s that share the same inner value (e.g. same invocation of [`Debug`]) will
+    /// return the same value.
+    pub fn get_mut<U: 'static>(&mut self) -> Option<&mut U> {
+        self.parent
+            .storage
+            .get_mut(&TypeId::of::<T>())?
+            .get_mut(&TypeId::of::<U>())?
+            .downcast_mut()
+    }
+
+    /// Insert a new value of type `U` into the storage of [`HookContext`].
+    ///
+    /// The returned value will the previously stored value of the same type `U` scoped over type
+    /// `T`, if it existed, did no such value exist it will return [`None`].
+    pub fn insert<U: 'static>(&mut self, value: U) -> Option<U> {
+        self.parent
+            .storage
+            .entry(TypeId::of::<T>())
+            .or_default()
+            .insert(TypeId::of::<U>(), Box::new(value))?
+            .downcast()
+            .map(|boxed| *boxed)
+            .ok()
+    }
+
+    /// Remove the value of type `U` from the storage of [`HookContext`] if it existed.
+    ///
+    /// The returned value will be the previously stored value of the same type `U`.
+    pub fn remove<U: 'static>(&mut self) -> Option<U> {
+        self.parent
+            .storage
+            .get_mut(&TypeId::of::<T>())?
+            .remove(&TypeId::of::<U>())?
+            .downcast()
+            .map(|boxed| *boxed)
+            .ok()
+    }
+
+    /// One of the most common interactions with [`HookContext`] is a counter to reference previous
+    /// frames or the content emitted during [`add_snippet()`].
+    ///
+    /// This is a utility method, which uses the other primitive methods provided to automatically
+    /// increment a counter, if the counter wasn't initialized this method will return `0`.
+    ///
+    /// ```rust
+    /// use std::io::ErrorKind;
+    ///
+    /// use error_stack::fmt::Emit;
+    /// use error_stack::Report;
+    ///
+    ///
+    /// Report::install_debug_hook::<()>(|_, ctx| {
```

Comment:
Can we use another type than `()` for the examples? This may teach the library wrong and people will use it regularly while this is not really expected but on types like `Suggestion`. Actually, `Suggestion` could be a good use case 🤷🏻‍♂️